packages/contracts/contracts/libraries/LazyMerkleTree.sol

```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import 'poseidon-solidity/PoseidonT3.sol';
+
+struct LazyTreeData {
+    uint32 maxIndex;
+    uint40 numberOfLeaves;
+    mapping(uint256 => uint256) elements;
+}
+
+library LazyMerkleTree {
+    uint256 internal constant SNARK_SCALAR_FIELD =
+        21888242871839275222246405745257275088548364400416034343698204186575808495617;
+    uint8 internal constant MAX_DEPTH = 32;
+    uint40 internal constant MAX_INDEX = (1 << 32) - 1;
+
+    function init(LazyTreeData storage self, uint8 depth) public {
+        require(depth <= MAX_DEPTH, 'LazyMerkleTree: Tree too large');
+        self.maxIndex = uint32((1 << depth) - 1);
+        self.numberOfLeaves = 0;
+    }
+
+    function reset(LazyTreeData storage self) public {
+        self.numberOfLeaves = 0;
+    }
+
+    function indexForElement(
+        uint8 level,
+        uint40 index
+    ) public pure returns (uint40) {
+        // store the elements sparsely
+        return MAX_INDEX * level + index;
+    }
+
+    function insert(LazyTreeData storage self, uint256 leaf) public {
+        uint40 index = self.numberOfLeaves;
+        self.numberOfLeaves = index + 1;
+        require(
+            leaf < SNARK_SCALAR_FIELD,
+            'LazyMerkleTree: leaf must be < SNARK_SCALAR_FIELD'
+        );
+        require(index < self.maxIndex, 'LazyMerkleTree: tree is full');
+
+        uint256 hash = leaf;
+
+        for (uint8 i = 0; true; ) {
+            self.elements[indexForElement(i, index)] = hash;
+            // it's a left element so we don't hash until there's a right element
+            if (index & 1 == 0) break;
+            uint40 elementIndex = indexForElement(i, index - 1);
+            hash = PoseidonT3.hash([self.elements[elementIndex], hash]);
+            unchecked {
+                index >>= 1;
+                i++;
+            }
+        }
+    }
+
+    function root(
+        LazyTreeData storage self,
+        uint8 depth
+    ) public view returns (uint256) {
+        // this will always short circuit if self.numberOfLeaves == 0
+        uint40 index = self.numberOfLeaves - 1;
+
+        uint256[MAX_DEPTH + 1] memory levels;
+
+        if (index & 1 == 0) {
+            levels[0] = self.elements[indexForElement(0, index)];
+        } else {
+            levels[0] = defaultZero(0);
+        }
+
+        for (uint8 i = 0; i < depth; ) {
+            if (index & 1 == 0) {
+                levels[i + 1] = PoseidonT3.hash([levels[i], defaultZero(i)]);
+            } else {
+                // if the element is a right sibling the parent is guaranteed to be calculated
+                uint256 parent = self.elements[
+                    indexForElement(i + 1, index >> 1)
+                ];
+                if (parent != 0) {
```

Comment:
Don't rely on a zero check for the parent if the tree is reusable. Instead track how many elements have been inserted on each level.